File: vlog/src/main/java/com/android/girish/vlog/chatheads/chatheads/VlogAdapter.java
package com.android.girish.vlog.chatheads.chatheads;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.android.girish.vlog.R;

import java.util.ArrayList;
import java.util.List;

public class VlogAdapter extends RecyclerView.Adapter<VlogAdapter.VlogViewHolder> {

    private static final String TAG = VlogAdapter.class.getSimpleName();
    private List<VlogModel> mFilteredLogList;
    private VlogModel mExpandedModel = null;

    public VlogAdapter() {
        mFilteredLogList = new ArrayList<>();
    }

    @Override
    public VlogViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new VlogViewHolder(LayoutInflater.from(parent.getContext()).inflate(
                R.layout.list_item_log, parent, false));
    }

    @Override
    public void onBindViewHolder(final VlogViewHolder holder, final int position) {
        final VlogModel model = mFilteredLogList.get(position);

        int priority = model.getLogPriority();
        switch (priority) {
            case VlogModel.ERROR:
                holder.logTag.setTextColor(Color.RED);
                holder.logMessage.setTextColor(Color.RED);
                break;
            case VlogModel.WARN:
                holder.logTag.setTextColor(Color.parseColor("#ff9966"));
                holder.logMessage.setTextColor(Color.parseColor("#ff9966"));
                break;
            default:
                holder.logTag.setTextColor(Color.BLACK);
                holder.logMessage.setTextColor(Color.BLACK);
                break;
        }
        holder.logTag.setText(getLogPriorityInitials(model.getLogPriority()) + "/" + model.getTag() + ": ");
        final boolean isExpanded = model == mExpandedModel;
        holder.logMessage.setText(isExpanded ? model.getLogMessage() : model.getLogMessage().length() > 30
                ? model.getLogMessage().substring(0, 28) : model.getLogMessage());
        holder.expandCollapseArrow.setImageResource(isExpanded ? R.drawable.ic_arrow_up : R.drawable.ic_arrow_down);
        holder.expandCollapseArrow.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mExpandedModel = isExpanded ? null : model;
                // TODO: @girish optimize if required -> update the item rather than the whole list.
                notifyDataSetChanged();
            }
        });
    }

    private String getLogPriorityInitials(int logPriority) {
        switch (logPriority) {
            case VlogModel.ASSERT:
                return "A";
            case VlogModel.DEBUG:
                return "D";
            case VlogModel.ERROR:
                return "E";
            case VlogModel.INFO:
                return "I";
            case VlogModel.VERBOSE:
                return "V";
            case VlogModel.WARN:
                return "W";
            default:
                return "";
        }
    }

    @Override
    public int getItemCount() {
        return mFilteredLogList != null ? mFilteredLogList.size() : 0;
    }

    public class VlogViewHolder extends RecyclerView.ViewHolder {


        TextView logTag;
        TextView logMessage;
        ImageView expandCollapseArrow;

        public VlogViewHolder(View itemView) {
            super(itemView);
            logTag = itemView.findViewById(R.id.log_tag);
            logMessage = itemView.findViewById(R.id.log_message);
            expandCollapseArrow = itemView.findViewById(R.id.arrow_img);
        }
    }

    public void addLogs(List<VlogModel> logs) {
        mFilteredLogList = logs;
        notifyDataSetChanged();
    }
}

File: vlog/src/main/java/com/android/girish/vlog/VlogAdapter.kt
package com.android.girish.vlog

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.android.girish.vlog.VlogAdapter.VlogViewHolder

internal class VlogAdapter : RecyclerView.Adapter<VlogViewHolder>() {
    private var mFilteredLogList: List<VlogModel>?
    private var mExpandedModel: VlogModel? = null
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VlogViewHolder {
        return VlogViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.list_item_log,
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: VlogViewHolder, position: Int) {
        val model = mFilteredLogList!![position]
        val priority = model.logPriority
        when (priority) {
            VlogModel.ERROR -> {
                holder.logTag.setTextColor(Color.RED)
                holder.logMessage.setTextColor(Color.RED)
            }
            VlogModel.WARN -> {
                holder.logTag.setTextColor(Color.parseColor("#ff9966"))
                holder.logMessage.setTextColor(Color.parseColor("#ff9966"))
            }
            else -> {
                holder.logTag.setTextColor(Color.BLACK)
                holder.logMessage.setTextColor(Color.BLACK)
            }
        }
        holder.logTag.text = getLogPriorityInitials(model.logPriority) + "/" + model.tag + ": "
        val isExpanded = model == mExpandedModel
        holder.logMessage.text = if (isExpanded) model.logMessage else if (model.logMessage.length > 30) model.logMessage.substring(0, 28) else model.logMessage
        holder.expandCollapseArrow.setImageResource(if (isExpanded) R.drawable.ic_arrow_up else R.drawable.ic_arrow_down)
        holder.expandCollapseArrow.setOnClickListener {
            mExpandedModel = if (isExpanded) null else model
            // TODO: @girish optimize if required -> update the item rather than the whole list.
            notifyDataSetChanged()
        }
    }

    private fun getLogPriorityInitials(logPriority: Int): String {
        return when (logPriority) {
            VlogModel.DEBUG -> "D"
            VlogModel.ERROR -> "E"
            VlogModel.INFO -> "I"
            VlogModel.VERBOSE -> "V"
            VlogModel.WARN -> "W"
            else -> ""
        }
    }

    override fun getItemCount(): Int {
        return if (mFilteredLogList != null) mFilteredLogList!!.size else 0
    }

    inner class VlogViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var logTag: TextView
        var logMessage: TextView
        var expandCollapseArrow: ImageView

        init {
            logTag = itemView.findViewById(R.id.log_tag)
            logMessage = itemView.findViewById(R.id.log_message)
            expandCollapseArrow = itemView.findViewById(R.id.arrow_img)
        }
    }

    fun addLogs(logs: List<VlogModel>?) {
        mFilteredLogList = logs
        notifyDataSetChanged()
    }

    companion object {
        private val TAG = VlogAdapter::class.java.simpleName
    }

    init {
        mFilteredLogList = ArrayList()
    }
}
